### PR TITLE
Add ERC721 example

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Add ERC721 example depends on PR https://github.com/Concordium/concordium-contracts-common/pull/22

## Changes

- Add ERC721 smart contract example.
- Update examples README.
- Export `HashMap`/`HashSet` from `contracts_common` in a collections module.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/erc721/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+
+    let parameter = SafeTransferFromParams {
+        from:         Address::Contract(ctx.self_address()),
+        to:           Address::Account(ctx.invoker()),
+        token_id:     params.token_id,
+        receive_name: None,
+        data:         vec![],
+    };
+
+    let mut receive_name_string = params.contract_name.get_chain_name().clone();
```
Comment (abizjak) [inline examples/erc721/src/lib.rs:468]: Shouldn't this be `contract_name` instead of `get_chain_name`? `chain_name` will have the `init_` prefix.

Comment (limemloh) [reply examples/erc721/src/lib.rs:468]: You are right

File: examples/erc721/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+ * of the owner.
+ * Note: This can be used to implement 'atomic swap' in another contract.
+ *
+ * ## Differences from the specification
+ *
+ * Since the specification is written to target the Ethereum blockchain, some
+ * details are implemented differently.
+ *
+ * - Instead of the getter functions the information should be read from the
+ *   contract state off-chain.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply examples/erc721/src/lib.rs]: I have removed this, and written a proper explanation

File: examples/erc721/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+ * - The specification uses a "zero address" (which is a special address used
+ *   as a null-address) which is not a thing on Concordium blockchain,
+ *   instead the address is wrapped in an Option if relevant.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply examples/erc721/src/lib.rs]: I have added a bit more detail, let me know if this is not enough

File: examples/erc721/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+    collections::{HashMap as Map, HashSet as Set},
+    *,
+};
+
+// Types
+
+/// Token Identifier, which combined with the address of the contract instance,
+/// forms the unique identifier of an NFT.
+/// Note: The specification requires a u256, but for efficiency we use u64,
+/// which is expected to be sufficient for most cases.
```
[the remark this answers is not in the document]
Comment (limemloh) [reply examples/erc721/src/lib.rs]: I have changed the phrasing to be a bit less vague

File: concordium-std/src/impls.rs [hunk 5]
```diff
@@ -11,6 +11,13 @@ use crate::{
 use concordium_contracts_common::*;
 use mem::MaybeUninit;
 
+#[cfg(not(feature = "std"))]
+use core::hash;
+#[cfg(feature = "std")]
+use std::hash;
```
Comment (abizjak) [inline concordium-std/src/impls.rs:17]: Distinguishing imports based on features should generally be done in `lib.rs` so as to not pollute the other files. The imports can be re-exported either with pub or pub(crate)

File: concordium-std/src/impls.rs [hunk 6]
```diff
@@ -734,6 +741,50 @@ impl<K: Deserial + Ord + Copy, V: Deserial> DeserialCtx for BTreeMap<K, V> {
     }
 }
 
+impl<K: Serial> SerialCtx for HashSet<K> {
```
Comment (abizjak) [inline concordium-std/src/impls.rs:744]: Document that the order is not specified.

File: concordium-std/src/impls.rs [hunk 7]
```diff
@@ -734,6 +741,50 @@ impl<K: Deserial + Ord + Copy, V: Deserial> DeserialCtx for BTreeMap<K, V> {
...
+        &self,
+        size_len: schema::SizeLength,
+        out: &mut W,
+    ) -> Result<(), W::Err> {
+        schema::serial_length(self.len(), size_len, out)?;
+        serial_hashset_no_length(self, out)
+    }
+}
+
+impl<K: Deserial + Eq + Hash> DeserialCtx for HashSet<K> {
```
Comment (abizjak) [inline concordium-std/src/impls.rs:755]: Document that the order is not specified.

File: concordium-std/src/impls.rs [hunk 8]
```diff
@@ -734,6 +741,50 @@ impl<K: Deserial + Ord + Copy, V: Deserial> DeserialCtx for BTreeMap<K, V> {
...
+        size_len: schema::SizeLength,
+        _ensure_ordered: bool,
+        source: &mut R,
+    ) -> ParseResult<Self> {
+        let len = schema::deserial_length(source, size_len)?;
+        deserial_hashset_no_length(source, len)
+    }
+}
+
+impl<K: Serial, V: Serial> SerialCtx for HashMap<K, V> {
```
Comment (abizjak) [inline concordium-std/src/impls.rs:766]: Document that the order is not specified.

File: concordium-std/src/impls.rs [hunk 9]
```diff
@@ -734,6 +741,50 @@ impl<K: Deserial + Ord + Copy, V: Deserial> DeserialCtx for BTreeMap<K, V> {
...
+        &self,
+        size_len: schema::SizeLength,
+        out: &mut W,
+    ) -> Result<(), W::Err> {
+        schema::serial_length(self.len(), size_len, out)?;
+        serial_hashmap_no_length(self, out)
+    }
+}
+
+impl<K: Deserial + Eq + Hash, V: Deserial> DeserialCtx for HashMap<K, V> {
```
Comment (abizjak) [inline concordium-std/src/impls.rs:777]: Document that the order is not specified.

Comment (abizjak) [reply concordium-std/src/impls.rs:777]: And that the ordered flag is meaningless.

**Review by abizjak — COMMENTED** [review]
As an example I think this is OK (after you add a bit more documentation to the serialization implementations).

But for a template/specification we'll need something a bit more direct.

**Review by abizjak — APPROVED** [review]

**Pull request merged**